packages/vite/src/node/plugins/asset.ts

```diff
@@ -344,18 +344,25 @@ async function fileToBuiltUrl(
   const file = cleanUrl(id)
   const content = await fsp.readFile(file)
 
-  if (shouldInline == null) {
-    shouldInline =
-      !!config.build.lib ||
-      // Don't inline SVG with fragments, as they are meant to be reused
-      (!(file.endsWith('.svg') && id.includes('#')) &&
-        !file.endsWith('.html') &&
-        content.length < Number(config.build.assetsInlineLimit) &&
-        !isGitLfsPlaceholder(content))
-  }
+  const finalShouldInline = (() => {
+    if (config.build.lib) return true
+    if (shouldInline != null) return shouldInline
+    let limit: number
+    if (typeof config.build.assetsInlineLimit === 'function') {
+      const userShouldInline = config.build.assetsInlineLimit(file, content)
+      if (userShouldInline != null) return userShouldInline
+      limit = DEFAULT_ASSETS_INLINE_LIMIT
+    } else {
+      limit = Number(config.build.assetsInlineLimit)
+    }
+    if (file.endsWith('.html')) return false
+    // Don't inline SVG with fragments, as they are meant to be reused
+    if (file.endsWith('.svg') && id.includes('#')) return false
+    return content.length < limit && !isGitLfsPlaceholder(content)
+  })()
```

Comment:
I guess it doesn't make a big difference, but creating this temporary function every time has a small cost. Extracting this might be better.